File: PROGRAMMERS/소수찾기.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

bool primeNumber(int k){
    if(k<2) return false;
    for(int i=2; i < k; i++){
        if(k % i ==0) return false;
    }
    return true;
}

int solution(string numbers) {
    int answer = 0;
    vector<int> v;
    for(int i=0; i<numbers.size(); i++){
        v.push_back(numbers[i]-'0');
    }
    sort(v.begin(), v.end());
    vector<int> a;
    
    //모든 경우의 수
    do{
        for(int i=0; i<=v.size(); i++){
            int val = 0;
            for(int j=0; j<i; j++){
                val = (val*10) + v[j];
                a.push_back(val);
            }
        }    
    }while(next_permutation(v.begin(), v.end()));

    //중복제거
    sort(a.begin(), a.end());
    a.erase(unique(a.begin(), a.end()), a.end());
       
    for(int i=0; i<a.size();i++){
        if(primeNumber(a[i]))
            answer ++;
    }

    return answer;
}

File: PROGRAMMERS/JAVA/소수찾기.java
import java.util.*;

class Solution {
    public int solution(int n) {
        ArrayList<Integer> list = new ArrayList<Integer>();
		
		int answer = 1;
		if(n >= 7) answer += 3;
		else if(n >= 5) answer += 2;
		else if(n >= 3) answer++;	
		
		for(int i=2; i<=n; i++) {
			
			if(i % 2 == 0) continue;
			else if(i % 3 == 0) continue;
			else if(i % 5 == 0) continue;
			else if(i % 7 == 0) continue;
			
			boolean q = false;
			if(i >= 121) {
				for(int p=0; p<list.size(); p++) {
					if(i < list.get(p)*list.get(p)) break;
					if(i % list.get(p) == 0) {
						q = true;
						break;
					}
				}
			}
			if(q) continue;
			answer++;
			list.add(i);

		}
		return answer;
    }
}